### ISSUE TITLE
TarCreationError <encodeOctal>: Tar value overflow

### ISSUE THREAD

**Issue by koterpillar** [fpco/cache-s3#1]
Ran `cache-s3 save stack` on a local machine, got the following:

```
[cache-s3][Info ][Mon, 29 Jan 2018 04:31:36 UTC]: Caching: <snip>/stack
cache-s3: TarCreationError "<encodeOctal>: Tar value overflow: (\"6423500\",Just 864)"
```

OS: macOS 10.13.3 (17D47)
cache-s3 version: unpacked from Github release `cache-s3-v0.1.1-osx-x86_64.tar.gz`

There might well be a lot of cruft in `.stack` because this is a development machine but I'm not sure how to figure out which file(s) are tripping `cache-s3`.

**Comment by lehins** [comment]
@koterpillar Thank you for reporting this issue. Implementation of offending function `encodeOctal` has been changed since v0.1.1 was released, so there is a chance that the error might have been fixed. This is definitely not enough information to pinpoint the exact problem, so I updated `tar-conduit` library by improving error reporting, so it should tell you which file and the exact tar header field that is triggering the problem. Could you please give [v0.1.1a](https://github.com/fpco/cache-s3/releases/tag/v0.1.1a) release a shot and see what will the error be? And maybe uploading an offending file somewhere, so I can replicate the issue.

**Comment by koterpillar** [comment]
```
[cache-s3][Info ][Mon, 29 Jan 2018 21:20:41 UTC]: Caching: /Users/xxxxxxxxxx/Projects/desktop/dotfiles/stack
cache-s3: TarCreationError "<encodeOctal>: Tar value overflow for file: /Users/xxxxxxxxxx/Projects/desktop/dotfiles/stack (for field 'ownerId' with maxLen 7): 1811961510"
```

Since length of _something_ is involved, I'll obfuscate sensitive values with `XXXX` but keep the same number of characters.

```
$ ls -l ~/.stack
lrwxr-xr-x 1 xxxxxxxxxx XXXX\Domain Users 49 Oct 25 10:37 /Users/xxxxxxxxxx/.stack -> /Users/xxxxxxxxxx/Projects/desktop/dotfiles/stack
$ ls -ld /Users/xxxxxxxxxx/Projects/desktop/dotfiles/stack
drwxr-xr-x 18 xxxxxxxxxx XXXX\Domain Users 576 Jan 29 16:55 /Users/xxxxxxxxxx/Projects/desktop/dotfiles/stack
```

**Comment by lehins** [comment]
@koterpillar Form the error it looks like we are hitting tar format userid/ownerid size limitation:
> POSIX 1988 'ustar' format is defined with *fixed-size* fields. There
is notably a 21 bits limit (2097151) for the UID and the GID.

As it seems, on MacOS it is not uncommon to see pretty large user ids, in your case it is `1811961510`, which simply doesn't fit in the tar header. That being said POSIX.1-2001 tar standard does handle arbitrary length user and group ids, but it hasn't been implemented yet.

**Comment by lehins** [comment]
@koterpillar Unfortunately I don't have access to a Mac, so I can't suggest a good workaround for this problem, unless maybe manually changing file ownership to a different user prior to saving the cache and changing it back after files get restored :/ So, I'll leave this ticket opened until snoyberg/tar-conduit#13 gets implemented, which I honestly don't have any ETA on.

**Comment by koterpillar** [comment]
Specifically for cache-s3, I don't see why the owner information has to be saved or restored at all, especially since the user IDs on different build agents might not match.

**Comment by lehins** [comment]
@koterpillar I do agree with you that restoring ownership isn't important, and possibly even the opposite from desired with this tool, since it would likely require root permissions to do so and there is a chance that user ID has changed between CI runs anyways. Despite those fact this error happened during tar file creation, rather than restoration and without proper support of large numbers in tar header this is not gonna fly. Defaulting to zero of course could be a sensible default, but instead I did go ahead and implement GNU style support for large numbers, so this issue should no longer bother you or any other mac user (Linux is affected too, but large user ids aren't so common). 
This is another prerelease: https://github.com/fpco/cache-s3/releases/tag/v0.1.2alpha that has described functionality included. Please let me know if it works for you.
Also thank you for the bug reports, I really appreciate it.

**Comment by koterpillar** [comment]
The new release works on Mac, I don't get this error anymore.

**Comment by lehins** [comment]
Fixed in [v0.1.2](https://github.com/fpco/cache-s3/releases/tag/v0.1.2)
